File: alchem-link/src/alchem_link/networks.py
"""Supported networks and how to reach them.

Every network here carries three things: the Alchemy subdomain used to build an
authenticated endpoint, a keyless public endpoint so the toolkit works before you have
an API key, and the chain id to verify you actually reached the chain you asked for.

Chain ids below were read back from the live endpoints, not copied from a table.
"""
from __future__ import annotations

import os
from dataclasses import dataclass
from typing import Dict, List, Optional

ALCHEMY_KEY_ENV = "ALCHEMY_API_KEY"
ALCHEMY_URL_ENV = "ALCHEMY_URL"


@dataclass(frozen=True)
class Network:
    key: str
    label: str
    chain_id: int
    native_symbol: str
    #: Alchemy hostname prefix — the endpoint is https://<subdomain>.g.alchemy.com/v2/<key>
    alchemy_subdomain: str
    #: Keyless fallback so `alchem-link price ETH/USD` works with zero setup.
    public_rpc: str
    explorer: str
    #: Testnet feeds carry test data from a separate node set. They must never be mixed
    #: into a cross-chain price comparison with mainnets — the numbers are unrelated, and
    #: averaging them in produces a consensus that describes nothing.
    testnet: bool = False
    #: Optimistic/ZK rollups need a sequencer-uptime check alongside any price read.
    #: See :mod:`alchem_link.sequencer`.
    layer2: bool = False

    def alchemy_url(self, api_key: str) -> str:
        return f"https://{self.alchemy_subdomain}.g.alchemy.com/v2/{api_key}"


NETWORKS: Dict[str, Network] = {
    "ethereum": Network(
        key="ethereum",
        label="Ethereum Mainnet",
        chain_id=1,
        native_symbol="ETH",
        alchemy_subdomain="eth-mainnet",
        public_rpc="https://ethereum-rpc.publicnode.com",
        explorer="https://etherscan.io",
    ),
    "sepolia": Network(
        key="sepolia",
        label="Ethereum Sepolia",
        chain_id=11155111,
        native_symbol="ETH",
        alchemy_subdomain="eth-sepolia",
        public_rpc="https://ethereum-sepolia-rpc.publicnode.com",
        explorer="https://sepolia.etherscan.io",
        testnet=True,
    ),
    "base": Network(
        key="base",
        label="Base Mainnet",
        chain_id=8453,
        native_symbol="ETH",
        alchemy_subdomain="base-mainnet",
        public_rpc="https://base-rpc.publicnode.com",
        explorer="https://basescan.org",
        layer2=True,
    ),
    "arbitrum": Network(
        key="arbitrum",
        label="Arbitrum One",
        chain_id=42161,
        native_symbol="ETH",
        alchemy_subdomain="arb-mainnet",
        public_rpc="https://arbitrum-one-rpc.publicnode.com",
        explorer="https://arbiscan.io",
        layer2=True,
    ),
    "optimism": Network(
        key="optimism",
        label="OP Mainnet",
        chain_id=10,
        native_symbol="ETH",
        alchemy_subdomain="opt-mainnet",
        public_rpc="https://optimism-rpc.publicnode.com",
        explorer="https://optimistic.etherscan.io",
        layer2=True,
    ),
    "polygon": Network(
        key="polygon",
        label="Polygon PoS",
        chain_id=137,
        native_symbol="POL",
        alchemy_subdomain="polygon-mainnet",
        public_rpc="https://polygon-bor-rpc.publicnode.com",
        explorer="https://polygonscan.com",
    ),
    "avalanche": Network(
        key="avalanche",
        label="Avalanche C-Chain",
        chain_id=43114,
        native_symbol="AVAX",
        alchemy_subdomain="avax-mainnet",
        public_rpc="https://avalanche-c-chain-rpc.publicnode.com",
        explorer="https://snowtrace.io",
    ),
    "bnb": Network(
        key="bnb",
        label="BNB Smart Chain",
        chain_id=56,
        native_symbol="BNB",
        alchemy_subdomain="bnb-mainnet",
        public_rpc="https://bsc-rpc.publicnode.com",
        explorer="https://bscscan.com",
    ),
    "gnosis": Network(
        key="gnosis",
        label="Gnosis Chain",
        chain_id=100,
        native_symbol="xDAI",
        alchemy_subdomain="gnosis-mainnet",
        public_rpc="https://gnosis-rpc.publicnode.com",
        explorer="https://gnosisscan.io",
    ),
    "scroll": Network(
        key="scroll",
        label="Scroll",
        chain_id=534352,
        native_symbol="ETH",
        alchemy_subdomain="scroll-mainnet",
        public_rpc="https://scroll-rpc.publicnode.com",
        explorer="https://scrollscan.com",
        layer2=True,
    ),
    "linea": Network(
        key="linea",
        label="Linea",
        chain_id=59144,
        native_symbol="ETH",
        alchemy_subdomain="linea-mainnet",
        public_rpc="https://linea-rpc.publicnode.com",
        explorer="https://lineascan.build",
        layer2=True,
    ),
}

DEFAULT_NETWORK = "ethereum"


def list_networks() -> List[Network]:
    return list(NETWORKS.values())


def get_network(key: str) -> Network:
    try:
        return NETWORKS[key.lower()]
    except KeyError:
        known = ", ".join(sorted(NETWORKS))
        raise KeyError(f"unknown network '{key}'. Known networks: {known}") from None


@dataclass(frozen=True)
class Endpoint:
    url: str
    #: Where the URL came from — shown by `doctor` so nobody debugs the wrong endpoint.
    source: str
    network: Network

    @property
    def is_authenticated(self) -> bool:
        return self.source in ("--rpc-url", ALCHEMY_URL_ENV, ALCHEMY_KEY_ENV)

    def redacted(self) -> str:
        """URL safe to print: an Alchemy key in the path is replaced with a placeholder."""
        marker = "/v2/"
        idx = self.url.find(marker)
        if idx == -1:
            return self.url
        return self.url[: idx + len(marker)] + "<key>"


def resolve_endpoint(
    network: str = DEFAULT_NETWORK,
    rpc_url: Optional[str] = None,
    env: Optional[Dict[str, str]] = None,
) -> Endpoint:
    """Pick the RPC endpoint to use, most explicit source first.

    1. ``--rpc-url`` / the ``rpc_url`` argument
    2. ``ALCHEMY_URL`` — a full endpoint, for anyone using a non-default host
    3. ``ALCHEMY_API_KEY`` — combined with the network's Alchemy subdomain
    4. the network's keyless public endpoint

    The fallback is what makes the toolkit usable in the first minute. It is rate
    limited and unauthenticated; `doctor` says so rather than letting you discover it
    under load.
    """
    net = get_network(network)
    environ = os.environ if env is None else env

    if rpc_url:
        return Endpoint(url=rpc_url, source="--rpc-url", network=net)

    explicit_url = environ.get(ALCHEMY_URL_ENV, "").strip()
    if explicit_url:
        return Endpoint(url=explicit_url, source=ALCHEMY_URL_ENV, network=net)

    api_key = environ.get(ALCHEMY_KEY_ENV, "").strip()
    if api_key:
        return Endpoint(url=net.alchemy_url(api_key), source=ALCHEMY_KEY_ENV, network=net)

    return Endpoint(url=net.public_rpc, source="public fallback", network=net)

File: web/lib/alchem/networks.ts
// Supported networks — a port of the table in `alchem_link/networks.py`.
//
// Static data only, so the client network picker can import it. Endpoint *resolution*
// lives in `endpoint.ts` and is server-only: it reads ALCHEMY_API_KEY, which must never
// reach the browser bundle. (Reading these endpoints from the page would fail anyway —
// most public RPC hosts send no CORS headers.)
//
// Every chain id below was read back from the live endpoint, not copied from a table.

export interface Network {
  key: string
  label: string
  chainId: number
  nativeSymbol: string
  /** Endpoint is https://<subdomain>.g.alchemy.com/v2/<key> */
  alchemySubdomain: string
  /** Keyless fallback, so the page works with zero setup. */
  publicRpc: string
  explorer: string
  /**
   * Testnet feeds carry test data from a separate node set. They must never be mixed
   * into a cross-chain price comparison with mainnets — the numbers are unrelated.
   */
  testnet?: boolean
  /**
   * Rollups need a sequencer-uptime check alongside any price read: the feed keeps
   * answering while the sequencer is down, with a price frozen at the moment it stopped.
   */
  layer2?: boolean
}

export const NETWORKS: Record<string, Network> = {
  ethereum: {
    key: 'ethereum',
    label: 'Ethereum Mainnet',
    chainId: 1,
    nativeSymbol: 'ETH',
    alchemySubdomain: 'eth-mainnet',
    publicRpc: 'https://ethereum-rpc.publicnode.com',
    explorer: 'https://etherscan.io',
  },
  sepolia: {
    key: 'sepolia',
    label: 'Ethereum Sepolia',
    chainId: 11155111,
    nativeSymbol: 'ETH',
    alchemySubdomain: 'eth-sepolia',
    publicRpc: 'https://ethereum-sepolia-rpc.publicnode.com',
    explorer: 'https://sepolia.etherscan.io',
    testnet: true,
  },
  base: {
    key: 'base',
    label: 'Base Mainnet',
    chainId: 8453,
    nativeSymbol: 'ETH',
    alchemySubdomain: 'base-mainnet',
    publicRpc: 'https://base-rpc.publicnode.com',
    explorer: 'https://basescan.org',
    layer2: true,
  },
  arbitrum: {
    key: 'arbitrum',
    label: 'Arbitrum One',
    chainId: 42161,
    nativeSymbol: 'ETH',
    alchemySubdomain: 'arb-mainnet',
    publicRpc: 'https://arbitrum-one-rpc.publicnode.com',
    explorer: 'https://arbiscan.io',
    layer2: true,
  },
  optimism: {
    key: 'optimism',
    label: 'OP Mainnet',
    chainId: 10,
    nativeSymbol: 'ETH',
    alchemySubdomain: 'opt-mainnet',
    publicRpc: 'https://optimism-rpc.publicnode.com',
    explorer: 'https://optimistic.etherscan.io',
    layer2: true,
  },
  polygon: {
    key: 'polygon',
    label: 'Polygon PoS',
    chainId: 137,
    nativeSymbol: 'POL',
    alchemySubdomain: 'polygon-mainnet',
    publicRpc: 'https://polygon-bor-rpc.publicnode.com',
    explorer: 'https://polygonscan.com',
  },
  avalanche: {
    key: 'avalanche',
    label: 'Avalanche C-Chain',
    chainId: 43114,
    nativeSymbol: 'AVAX',
    alchemySubdomain: 'avax-mainnet',
    publicRpc: 'https://avalanche-c-chain-rpc.publicnode.com',
    explorer: 'https://snowtrace.io',
  },
  bnb: {
    key: 'bnb',
    label: 'BNB Smart Chain',
    chainId: 56,
    nativeSymbol: 'BNB',
    alchemySubdomain: 'bnb-mainnet',
    publicRpc: 'https://bsc-rpc.publicnode.com',
    explorer: 'https://bscscan.com',
  },
  gnosis: {
    key: 'gnosis',
    label: 'Gnosis Chain',
    chainId: 100,
    nativeSymbol: 'xDAI',
    alchemySubdomain: 'gnosis-mainnet',
    publicRpc: 'https://gnosis-rpc.publicnode.com',
    explorer: 'https://gnosisscan.io',
  },
  scroll: {
    key: 'scroll',
    label: 'Scroll',
    chainId: 534352,
    nativeSymbol: 'ETH',
    alchemySubdomain: 'scroll-mainnet',
    publicRpc: 'https://scroll-rpc.publicnode.com',
    explorer: 'https://scrollscan.com',
    layer2: true,
  },
  linea: {
    key: 'linea',
    label: 'Linea',
    chainId: 59144,
    nativeSymbol: 'ETH',
    alchemySubdomain: 'linea-mainnet',
    publicRpc: 'https://linea-rpc.publicnode.com',
    explorer: 'https://lineascan.build',
    layer2: true,
  },
}

export const DEFAULT_NETWORK = 'ethereum'

export const NETWORK_KEYS = Object.keys(NETWORKS)

/**
 * Chainlink L2 sequencer uptime feeds — a port of `alchem_link/sequencer.py`.
 *
 * Each was verified live: all three report `L2 Sequencer Uptime Status Feed` as their
 * own `description()`.
 */
export const SEQUENCER_FEEDS: Record<string, string> = {
  arbitrum: '0xFdB631F5EE196F0ed6FAa767959853A9F217697D',
  optimism: '0x371EAD81c9102C9BF4874A9075FFFf170F2Ee389',
  base: '0xBCF85224fc0756B9Fa45aA7892530B47e10b6433',
}

/** Chainlink's documented grace period after a sequencer restart, in seconds. */
export const SEQUENCER_GRACE_PERIOD_SECS = 3600

export function listNetworks(): Network[] {
  return Object.values(NETWORKS)
}

export function getNetwork(key: string): Network {
  const net = NETWORKS[key.toLowerCase()]
  if (!net) {
    throw new Error(`unknown network '${key}'. Known networks: ${NETWORK_KEYS.slice().sort().join(', ')}`)
  }
  return net
}

// Endpoint resolution — `resolveEndpoint`, `redact`, and the `Endpoint` type — lives in
// `endpoint.ts`, which is server-only. Keeping it out of this file is what lets the
// client import the network table without dragging `process.env` along with it.
